### PR TITLE
Inject feature gate instance into client-go for kube components.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind feature
/sig api-machinery

#### What this PR does / why we need it:

Registers all client-go features with the DefaultFeatureGate instance and wires client-go to perform feature checks using the same instance used for kube feature checks. For kube components, this means that client features will be exposed in the existing `--feature-gates` flag.

In order to avoid a dependency cycle between component-base and client-go, client-go maintains parallel definitions of component-base's feature types and constants. Passing kube's default feature gate instance to client-go requires an adapter.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Proposal]: https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs/edit#heading=h.ookhkzz1jrsp
```

### CODE REVIEW

**Comment by benluddy** [comment]
/assign @p0lyn0mial 

/hold

Holding for proof PR demonstrating a client feature wired to a kube component.

File: pkg/features/client_adapter.go
Comment (p0lyn0mial) [inline pkg/features/client_adapter.go:18]: Would it make sense and if yes how hard it would be to write a unit test that would make sure no new field were added to `clientfeatures.FeatureSpec` ? (reflection?)

**Comment by p0lyn0mial** [comment]
> Holding for proof PR demonstrating a client feature wired to a kube component.

opened a proof PR https://github.com/kubernetes/kubernetes/pull/122608 for KCM

**Comment by p0lyn0mial** [comment]
@benluddy since https://github.com/kubernetes/kubernetes/pull/122571 merged you will need to slightly change [TestKubeFeatures](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features_test.go#L24C1-L24C1). 

I already did that in https://github.com/kubernetes/kubernetes/pull/122608. Feel free to pull https://github.com/kubernetes/kubernetes/pull/122608/commits/feaa22ece64893dbde3040a7d24b04bb70fb6239 in here.

File: pkg/features/client_adapter.go
Comment (benluddy) [reply pkg/features/client_adapter.go:10]: Renamed to `clientAdapter`. It adapts to both `clientfeatures.Gates` and `clientfeatures.Registry` and lives in a package named `features`, so I felt this was more reasonable than having separate adapters per client interface or naming the type based on only one of the implemented interfaces. WDYT?

File: pkg/features/client_adapter.go
Comment (benluddy) [reply pkg/features/client_adapter.go]: As discussed offline, we are expecting impactful mistakes to be caught by compilation failures or panics at package-initialization time and won't include a unit test.

**Comment by benluddy** [comment]
> @benluddy since #122571 merged you will need to slightly change [TestKubeFeatures](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features_test.go#L24C1-L24C1).
> 
> I already did that in #122608. Feel free to pull [feaa22e](https://github.com/kubernetes/kubernetes/commit/feaa22ece64893dbde3040a7d24b04bb70fb6239) in here.

I picked that commit into this PR and added a text description to the test cases to clarify the exact behaviors under test. Please double-check that my text is accurate.

File: pkg/features/kube_features.go
Comment (deads2k) [inline pkg/features/kube_features.go:909]: explain why, not just what.  This wires the kube binary CLI flags as the authoritative source of the client-go featuregates.

**Comment by deads2k** [comment]
/lgtm
/approve

@benluddy you may release hold when ready.

**Comment by benluddy** [comment]
/assign @dims 

PTAL on behalf of dep-approvers. This PR adds an import of `k8s.io/client-go/features` so there's a new corresponding line in `vendor/modules.txt`.

**Review by soltysh — APPROVED** [review]
/approve
deps

**Comment by benluddy** [comment]
> /test pull-kubernetes-verify

Looks like this was an example of https://github.com/kubernetes/kubernetes/issues/121865

**Pull request merged**
